pkg/segment/query/iqr/intermediateQueryResult.go

```diff
@@ -716,17 +724,151 @@ func (iqr *IQR) AsResult() (*structs.PipeSearchResponseOuter, error) {
+			Hits: structs.PipeSearchResponse{
+				TotalMatched: iqr.NumberOfRecords(),
+				Hits:         recordsAsAny,
+			},
+			AllPossibleColumns: toputils.GetKeysOfMap(records),
+			Errors:             nil,
+			Qtype:              iqr.qType.String(),
+			CanScrollMore:      false,
+			ColumnsOrder:       toputils.GetSortedStringKeys(records),
+		}
+	} else if iqr.qType == structs.GroupByCmd { // TODO: Handle SegmentStatsCmd
+		bucketHolderArr, aggGroupByCols, measureFuncs, bucketCount, err := iqr.getFinalResultForGroupBy()
+		if err != nil {
+			return nil, toputils.TeeErrorf("IQR.AsResult: error getting final result for GroupBy: %v", err)
+		}
+
+		response = &structs.PipeSearchResponseOuter{
+			Hits: structs.PipeSearchResponse{
+				TotalMatched: 0, // TODO: get the total matched records
+				Hits:         nil,
+			},
+			AllPossibleColumns: toputils.GetKeysOfMap(records),
+			Errors:             nil,
+			Qtype:              iqr.qType.String(),
+			CanScrollMore:      false,
+			ColumnsOrder:       toputils.GetSortedStringKeys(records),
+			BucketCount:        bucketCount,
+			MeasureFunctions:   measureFuncs,
+			MeasureResults:     bucketHolderArr,
+			GroupByCols:        aggGroupByCols,
+			// TODO: add IsTimechart flag
+		}
 	}
 
 	return response, nil
 }
+
+func (iqr *IQR) AppendGroupByResults(bucketHolderArr []*structs.BucketHolder, measureFuncs []string, aggGroupByCols []string, bucketCount int) error {
+	if err := iqr.validate(); err != nil {
+		log.Errorf("IQR.AppendGroupByResults: validation failed: %v", err)
+		return err
+	}
+
+	if iqr.qType != structs.GroupByCmd {
+		return fmt.Errorf("IQR.AppendGroupByResults: not a GroupBy query")
+	}
+
+	iqr.mode = withoutRRCs
+
+	knownValues := make(map[string][]utils.CValueEnclosure)
+
+	for _, aggGroupByCol := range aggGroupByCols {
+		knownValues[aggGroupByCol] = make([]utils.CValueEnclosure, len(bucketHolderArr))
+	}
+	for _, retMFun := range measureFuncs {
+		knownValues[retMFun] = make([]utils.CValueEnclosure, len(bucketHolderArr))
+	}
+
+	conversionErrors := make(map[string]interface{})
+
+	for i, bucketHolder := range bucketHolderArr {
+		for idx, aggGroupByCol := range aggGroupByCols {
+			colValue := bucketHolder.GroupByValues[idx]
+			err := knownValues[aggGroupByCol][i].ConvertValue(colValue)
+			if err != nil {
+				conversionErrors[fmt.Sprintf("%v_%v", i, aggGroupByCol)] = colValue
+			}
+		}
+
+		for _, measureFunc := range measureFuncs {
+			value := bucketHolder.MeasureVal[measureFunc]
+			err := knownValues[measureFunc][i].ConvertValue(value)
+			if err != nil {
+				conversionErrors[fmt.Sprintf("%v_%v", i, measureFunc)] = value
+			}
+		}
+	}
+
+	err := iqr.AppendKnownValues(knownValues)
+	if err != nil {
+		return err
+	}
+
+	iqr.groupbyColumns = append(iqr.groupbyColumns, aggGroupByCols...)
+	iqr.measureColumns = append(iqr.measureColumns, measureFuncs...)
+
+	if len(conversionErrors) > 0 {
+		log.Errorf("IQR.AppendGroupByResults: conversion errors: %v", conversionErrors)
```

Comment:
if there are indeed errors, then would the len of this conversionErrors be equal to the max buckets we support (in case all buckets had errors). 
To be safe, we should log upto a max of 5 elements of this array.